hudi-gcp/src/main/java/org/apache/hudi/gcp/bigquery/BigQuerySyncTool.java

```diff
@@ -52,34 +57,55 @@ public class BigQuerySyncTool extends HoodieSyncTool {
 
   private static final Logger LOG = LoggerFactory.getLogger(BigQuerySyncTool.class);
 
-  public final BigQuerySyncConfig config;
-  public final String tableName;
-  public final String manifestTableName;
-  public final String versionsTableName;
-  public final String snapshotViewName;
+  private final BigQuerySyncConfig config;
+  private final String tableName;
+  private final String manifestTableName;
+  private final String versionsTableName;
+  private final String snapshotViewName;
+  private final ManifestFileWriter manifestFileWriter;
+  private final HoodieBigQuerySyncClient bqSyncClient;
+  private final HoodieTableMetaClient metaClient;
+  private final BigQuerySchemaResolver bqSchemaResolver;
 
   public BigQuerySyncTool(Properties props) {
-    super(props);
+    // will build file writer, client, etc. from configs
+    this(props, null, null, null, null);
+  }
+
+  @VisibleForTesting // allows us to pass in mocks for the writer and client
+  BigQuerySyncTool(Properties properties, ManifestFileWriter manifestFileWriter, HoodieBigQuerySyncClient providedBqSyncClient, HoodieTableMetaClient providedMetaClient,
+                   BigQuerySchemaResolver providedBqSchemaResolver) {
+    super(properties);
     this.config = new BigQuerySyncConfig(props);
     this.tableName = config.getString(BIGQUERY_SYNC_TABLE_NAME);
     this.manifestTableName = tableName + "_manifest";
     this.versionsTableName = tableName + "_versions";
     this.snapshotViewName = tableName;
+    this.bqSyncClient = providedBqSyncClient == null ? new HoodieBigQuerySyncClient(config) : providedBqSyncClient;
+    // reuse existing meta client if not provided (only test cases will provide their own meta client)
```

Comment:
then, how do we ensure we can catch issues within buildManifestFileWriterFromConfig, HoodieBigQuerySyncClient() constructor etc, if we override all these w/ provided entities in tests?

Reply:
I also don't have a good way around this. my only concern is, if we start to set precedence, down the line everyone will start to follow similar pattern and might miss out on some actual issues in the source code. May be in this patch, I don't see major gaps. but once we set precedence, its hard to keep track (since we don't review every patch in OSS)

Reply:
See my above question, dependency injection is a common practice but Hudi makes it hard to follow this due to its constraints. Can we consider adopting some patterns that are more friendly to testing so we can adopt unit testing over functional testing in general?

Reply:
One idea is to put an extra layer between translating the props into the objects. So the BigQuerySyncTool itself wouldn't have any logic other than creating some objects passed into another class. That other class will then be easy to unit test but the translation of props to inputs to that class will still be untested.

Reply:
lets not block this PR on this discussion. we can take it async. lets go ahead w/ the patch. I am good